src/upload-button.tsx

```diff
@@ -0,0 +1,351 @@
+                        />
+                    );
+                } catch (exc) {
+                    resetInput();
+                    return;
+                }
+
+                if (resolution.skip) {
+                    if (resolution.applyToAll)
+                        skipAll = true;
+                    continue;
+                }
+
+                if (resolution.applyToAll && resolution.replace)
+                    replaceAll = true;
+
+                toUploadFiles.push(inputFile);
+            } else {
+                toUploadFiles.push(inputFile);
+            }
+        }
+
+        if (toUploadFiles.length === 0) {
+            resetInput();
+            return;
+        }
+
+        window.addEventListener("beforeunload", beforeUnloadHandler);
+
+        const cancelledUploads = [];
+        await Promise.allSettled(toUploadFiles.map(async (file: File) => {
+            const tmp_path = path.slice();
+            tmp_path.push(file.name);
+            const destination = tmp_path.join('/');
+            const abort = new AbortController();
+
+            setUploadedFiles(oldFiles => {
+                return {
+                    [file.name]: { file, progress: 0, cancel: () => abort.abort() },
+                    ...oldFiles,
+                };
+            });
+
+            try {
+                await upload(destination, file, (progress) => {
+                    const now = performance.now();
+                    if (now < next_progress)
+                        return;
+                    next_progress = now + 200; // only rerender every 200ms
+                    setUploadedFiles(oldFiles => {
+                        const oldFile = oldFiles[file.name];
+                        return {
+                            ...oldFiles,
+                            [file.name]: { ...oldFile, progress },
+                        };
+                    });
+                }, abort.signal);
+                // TODO: pass { superuser: try } depending on directory owner
+            } catch (exc) {
+                cockpit.assert(exc instanceof Error, "Unknown exception type");
+                if (exc instanceof DOMException && exc.name === 'AbortError') {
+                    addAlert(_("Cancelled"), AlertVariant.warning, "upload",
+                             cockpit.format(_("Cancelled upload of $0"), file.name));
+                } else {
+                    addAlert(_("Upload error"), AlertVariant.danger, "upload-error", exc.toString());
+                }
+                cancelledUploads.push(file);
+            } finally {
+                setUploadedFiles(oldFiles => {
+                    const copy = { ...oldFiles };
+                    delete copy[file.name];
+                    return copy;
+                });
+            }
+        }));
+
+        resetInput();
+        window.removeEventListener("beforeunload", beforeUnloadHandler);
```

Comment:
Note: was going to complain that this belongs in a `finally` block, but it seems that `allSettled` never throws.  Convenient.